propolis/src/block.rs

```diff
@@ -53,53 +54,121 @@ pub struct BlockInquiry {
 pub trait BlockDev<R: BlockReq>: Send + Sync + 'static {
     /// Enqueues a [`BlockReq`] to the underlying device.
     fn enqueue(&self, req: R);
+
     /// Requests metadata about the block device.
     fn inquire(&self) -> BlockInquiry;
 }
 
+/// Abstraction over an actual backing store
+pub trait BlockDevBackingStore: Send + Sync + 'static {
+    fn issue_read(&self, offset: usize, sz: usize, mapping: SubMapping) -> Result<usize>;
+    fn issue_write(&self, bytes: &[u8], offset: usize) -> Result<usize>;
+    fn issue_flush(&self) -> Result<()>;
+
+    fn is_ro(&self) -> bool;
+
+    // length in bytes
+    fn len(&self) -> usize;
+}
+
 /// A block device implementation backed by an underlying file.
-///
-/// Primarily accessed through the [`BlockDev`] interface.
-pub struct PlainBdev<R: BlockReq> {
+pub struct FileBlockDevBackingStore {
     fp: File,
     is_ro: bool,
-    is_raw: bool,
-    block_size: usize,
-    sectors: usize,
-    reqs: Mutex<VecDeque<R>>,
-    cond: Condvar,
+    len: usize,
 }
 
-impl<R: BlockReq> PlainBdev<R> {
-    /// Creates a new block device from a device at `path`.
-    pub fn create(path: impl AsRef<Path>, readonly: bool) -> Result<Arc<Self>> {
+impl FileBlockDevBackingStore {
+    pub fn from_path(
+        path: impl AsRef<Path>,
+        readonly: bool,
+    ) -> Result<FileBlockDevBackingStore> {
         let p: &Path = path.as_ref();
 
         let meta = metadata(p)?;
         let is_ro = readonly || meta.permissions().readonly();
 
         let fp = OpenOptions::new().read(true).write(!is_ro).open(p)?;
-        let is_raw = fp.metadata()?.file_type().is_char_device();
+        let len = fp.metadata().unwrap().len() as usize;
+
+        Ok(FileBlockDevBackingStore { fp, is_ro, len })
+    }
+
+    pub fn get_file(&self) -> &File {
+        &self.fp
+    }
+}
+
+impl BlockDevBackingStore for FileBlockDevBackingStore {
+    fn issue_read(&self, offset: usize, sz: usize, mapping: SubMapping) -> Result<usize> {
+        let nread = mapping.pread(&self.fp, sz, offset as i64)?;
+        Ok(nread as usize)
```

Comment:
The return types should match up so you can simplify this a bit:
```suggestion
        mapping.pread(&self.fp, sz, offset as i64)
```

Reply:
Fixed in 7d9e407